### PR TITLE
Opsrc 397/export selected to pdf tests

### PR DESCRIPTION
**Pull request by PiotrSzymanski2000**

### CODE REVIEW

File: features/adding_wishlist_product_to_cart.feature [hunk 1]
```diff
@@ -26,4 +27,37 @@ Feature: Adding wishlist product to cart
...
+        Then I should have "Bushmills Black Bush Whiskey" product in my cart
+
+
+
+
+
+
+
+
+
```
Comment (GracjanJozefczyk) [inline features/adding_wishlist_product_to_cart.feature:63]: Empty lines

File: features/removing_product_from_wishlist.feature [hunk 2]
```diff
@@ -15,3 +15,14 @@ Feature: Removing a product from wishlist
...
+    @ui
+    Scenario: Removing selected products from wishlist
+        Given the store has a product "Jack Daniels Gentleman" priced at "$10.00"
+        And I have this product in my wishlist
+        When I go to the wishlist page
+        Then I check "Jack Daniels Gentleman"
+        And I remove selected products from wishlist
+        And I should have 0 products in my wishlist
+
+
```
Comment (GracjanJozefczyk) [inline features/removing_product_from_wishlist.feature:28]: Empty line

File: spec/Controller/Action/ExportWishlistToPdfActionSpec.php [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        $wishlistProductCommand->setWishlistProduct($wishlistProduct);
+        $commandsArray->add($wishlistProductCommand);
+
+        $formFactory
+            ->create(
+                WishlistCollectionType::class,
+                ['items' => $commandsArray],
+                [
+                    'cart' => $cart,
+                ]
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ExportWishlistToPdfActionSpec.php:106]: One array is in single line, another multiline

File: spec/Controller/Action/ExportWishlistToPdfActionSpec.php [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ExportWishlistToPdfActionSpec.php:162]: One array is in single line, another multiline

File: spec/Controller/Action/ListWishlistProductsActionSpec.php [hunk 5]
```diff
@@ -63,32 +67,42 @@ function it_is_initializable(): void
...
         $formFactory
             ->create(
-                AddProductsToCartType::class,
-                null,
+                WishlistCollectionType::class,
+                ['items' => $commandsArray],
                 [
                     'cart' => $cart,
-                    'wishlist_products' => $wishlistProducts,
                 ]
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ListWishlistProductsActionSpec.php:102]: One array is in single line, another multiline

File: spec/Controller/Action/ListWishlistProductsActionSpec.php [hunk 6]
```diff
@@ -100,71 +114,84 @@ function it_lists_wishlist_items(
...
-        EntityManagerInterface $cartManager,
-        Environment $twigEnvironment,
-        Response $response
-    ): void {
+        OrderItemInterface        $cartItem,
+        OrderModifierInterface    $orderModifier,
+        EntityManagerInterface    $cartManager,
+        Environment               $twigEnvironment,
+        Response                  $response,
+        AddWishlistProductInterface        $wishlistProductsCommand
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ListWishlistProductsActionSpec.php:141]: Forgot about last one ;)

File: spec/Controller/Action/ListWishlistProductsActionSpec.php [hunk 7]
```diff
@@ -100,71 +114,84 @@ function it_lists_wishlist_items(
...
         $formFactory
             ->create(
-                AddProductsToCartType::class,
-                null,
+                WishlistCollectionType::class,
+                ['items' => $commandsArray],
                 [
                     'cart' => $cart,
-                    'wishlist_products' => $wishlistProducts,
                 ]
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ListWishlistProductsActionSpec.php:163]: One array is in single line, another multiline

File: src/Command/Wishlist/AddWishlistProduct.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
+<?php
+declare(strict_types=1);
```
Comment (GracjanJozefczyk) [inline src/Command/Wishlist/AddWishlistProduct.php:2]: No empty line, no BitBag Header

File: src/Exporter/ExporterWishlistToPdf.php [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        Environment $twigEnvironment
+    )
+    {
+        $this->productVariantRepository = $productVariantRepository;
+        $this->variantImagePathResolver = $variantImagePathResolver;
+        $this->variantPdfModelFactory = $variantPdfModelFactory;
+        $this->twigEnvironment = $twigEnvironment;
+    }
+
+
```
Comment (GracjanJozefczyk) [inline src/Exporter/ExporterWishlistToPdf.php:48]: Empty line

File: src/Exporter/ExporterWishlistToPdfInterface.php [hunk 10]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+/*
+ * This file was created by developers working at BitBag
+ * Do you need more information about us and what we do? Visit our https://bitbag.io website!
+ * We are hiring developers from all over the world. Join us and start your new, exciting adventure and become part of us: https://bitbag.io/career
+*/
+
+
```
Comment (GracjanJozefczyk) [inline src/Exporter/ExporterWishlistToPdfInterface.php:9]: Empty line

File: spec/Controller/Action/ExportWishlistToPdfActionSpec.php [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    function let(
+        WishlistContextInterface          $wishlistContext,
+        CartContextInterface              $cartContext,
+        FormFactoryInterface              $formFactory,
+        FlashBagInterface                 $flashBag,
+        TranslatorInterface               $translator,
+        UrlGeneratorInterface             $urlGenerator,
+        Environment                       $twigEnvironment,
+        ExporterWishlistToPdfInterface    $exporterWishlistToPdf
+    ): void
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ExportWishlistToPdfActionSpec.php:47]:
```suggestion
    ): void {
        $this->beConstructedWith(
```

File: spec/Controller/Action/ExportWishlistToPdfActionSpec.php [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        FormFactoryInterface                $formFactory,
+        FormInterface                       $form,
+        FormErrorIterator                   $formErrorIterator,
+        ExporterWishlistToPdfInterface      $exporterWishlistToPdf,
+        FlashBagInterface                   $flashBag,
+        TranslatorInterface                 $translator,
+        UrlGeneratorInterface               $urlGenerator,
+        ArrayCollection                     $arrayCollection
+
+    ):  void
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ExportWishlistToPdfActionSpec.php:81]:
```suggestion
    ):  void {
```

File: spec/Controller/Action/ExportWishlistToPdfActionSpec.php [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+        WishlistInterface                   $wishlist,
+        CartContextInterface                $cartContext,
+        OrderInterface                      $cart,
+        FormFactoryInterface                $formFactory,
+        FormInterface                       $form,
+        FormErrorIterator                   $formErrorIterator,
+        FormView                            $formView,
+        Environment                         $twigEnvironment
+
+    ):  void
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ExportWishlistToPdfActionSpec.php:138]:
```suggestion
    ):  void {
```

File: spec/Controller/Action/ExportWishlistToPdfActionSpec.php [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        $translator->trans('bitbag_sylius_wishlist_plugin.ui.select_products')->willReturn("Select at least one item.");
+        $flashBag->add('error', "Select at least one item.");
+        $urlGenerator
+            ->generate('bitbag_sylius_wishlist_plugin_shop_wishlist_list_products')
+            ->willReturn('Content');
+
+        $this->__invoke($request)->shouldHaveType(RedirectResponse::class);
+    }
+
+
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ExportWishlistToPdfActionSpec.php:125]: Empty line

File: spec/Controller/Action/ListWishlistProductsActionSpec.php [hunk 15]
```diff
@@ -36,14 +38,15 @@ final class ListWishlistProductsActionSpec extends ObjectBehavior
...
-        Environment $twigEnvironment
-    ): void {
+        CartContextInterface     $cartContext,
+        FormFactoryInterface     $formFactory,
+        OrderModifierInterface   $orderModifier,
+        EntityManagerInterface   $cartManager,
+        FlashBagInterface        $flashBag,
+        TranslatorInterface      $translator,
+        Environment              $twigEnvironment
+    ):  void
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ListWishlistProductsActionSpec.php:48]:
```suggestion
    ):  void {
```

File: spec/Controller/Action/ListWishlistProductsActionSpec.php [hunk 16]
```diff
@@ -63,32 +66,43 @@ function it_is_initializable(): void
...
+        WishlistInterface        $wishlist,
+        CartContextInterface     $cartContext,
+        OrderInterface           $cart,
+        Collection               $wishlistProducts,
+        FormFactoryInterface     $formFactory,
+        FormInterface            $form,
+        FormErrorIterator        $formErrorIterator,
+        FormView                 $formView,
+        Environment              $twigEnvironment
+    ):  void
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ListWishlistProductsActionSpec.php:79]:
```suggestion
    ):  void {
```

File: spec/Controller/Action/ListWishlistProductsActionSpec.php [hunk 17]
```diff
@@ -100,71 +114,84 @@ function it_lists_wishlist_items(
...
+        FormFactoryInterface            $formFactory,
+        FormInterface                   $form,
+        FormErrorIterator               $formErrorIterator,
+        FormView                        $formView,
+        OrderItemInterface              $cartItem,
+        OrderModifierInterface          $orderModifier,
+        EntityManagerInterface          $cartManager,
+        Environment                     $twigEnvironment,
+        AddWishlistProductInterface     $wishlistProductsCommand
+    ): void
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ListWishlistProductsActionSpec.php:140]:
```suggestion
    ): void {
```

File: spec/Controller/Action/ListWishlistProductsActionSpec.php [hunk 18]
```diff
@@ -100,71 +114,84 @@ function it_lists_wishlist_items(
...
-        $form->handleRequest($request)->shouldBeCalled();
-        $orderModifier->addToOrder($cart, $cartItem)->shouldBeCalled();
-        $cartManager->persist($cart)->shouldBeCalled();
         $cartManager->flush()->shouldBeCalled();
         $form->getErrors()->shouldNotBeCalled();
 
         $this->__invoke($request)->shouldHaveType(Response::class);
     }
-}
+}
```
Comment (GracjanJozefczyk) [inline spec/Controller/Action/ListWishlistProductsActionSpec.php:197]: Missing empty line

File: spec/Exporter/ExporterWishlistToPdfSpec.php [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+use Twig\Environment;
+
+final class ExporterWishlistToPdfSpec extends ObjectBehavior
+{
+    function let(
+        ProductVariantRepositoryInterface   $productVariantRepository,
+        VariantImagePathResolverInterface   $variantImagePathResolver,
+        VariantPdfModelFactoryInterface     $variantPdfModelFactory,
+        Environment                         $twigEnvironment
+    ): void
```
Comment (GracjanJozefczyk) [inline spec/Exporter/ExporterWishlistToPdfSpec.php:38]:
```suggestion
    ): void {
```

File: spec/Exporter/ExporterWishlistToPdfSpec.php [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+        $this->handleCartItems(
+            new ArrayCollection(
+                [
+                    $wishlistProduct->getWrappedObject()
+                ]
+            ),
+            $request
+        )->shouldReturn(true);
+    }
+}
```
Comment (GracjanJozefczyk) [inline spec/Exporter/ExporterWishlistToPdfSpec.php:124]: Missing empty line

File: spec/Exporter/ExporterWishlistToPdfSpec.php [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+        Request                             $request,
+        ProductVariantRepositoryInterface   $productVariantRepository,
+        ProductVariantInterface             $productVariant,
+        AddWishlistProductInterface         $wishlistProduct,
+        WishlistProductInterface            $product,
+        AddToCartCommandInterface           $addToCartCommand,
+        OrderItemInterface                  $orderItem,
+        VariantImagePathResolverInterface   $variantImagePathResolver,
+        VariantPdfModelFactoryInterface     $variantPdfModelFactory
+    ): void
```
Comment (GracjanJozefczyk) [inline spec/Exporter/ExporterWishlistToPdfSpec.php:93]:
```suggestion
    ): void {
```

File: spec/Exporter/ExporterWishlistToPdfSpec.php [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+        $addWishlistProduct->isSelected()->willReturn(false);
+        $this->handleCartItems($arrayCollection, new Request())->shouldReturn(false);
+    }
+
+    function it_throws_404_when_product_is_not_found(
+        Request                             $request,
+        ProductVariantRepositoryInterface   $productVariantRepository,
+        AddWishlistProductInterface         $wishlistProduct,
+        WishlistProductInterface            $product
+    ): void
```
Comment (GracjanJozefczyk) [inline spec/Exporter/ExporterWishlistToPdfSpec.php:71]:
```suggestion
    ): void {
```

File: spec/Model/Factory/VariantPdfModelFactorySpec.php [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+        $productPdfModel->getActualVariant()->shouldReturn('variant test');
+
+        $this->createWithVariantAndImagePath(
+            $productVariant,
+            'http://127.0.0.1:8000/media/image/b4/c2/fc6b3202ee567e0fb05f293b709c.jpg',
+            10,
+            'variant test'
+        )->shouldBeAnInstanceOf(VariantPdfModel::class);
+    }
+}
```
Comment (GracjanJozefczyk) [inline spec/Model/Factory/VariantPdfModelFactorySpec.php:54]: Missing empty line

File: spec/Model/VariantPdfModelSpec.php [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+        $this->setVariant($productVariant);
+        $this->setImagePath('/image/123/image.jpg');
+        $this->setQuantity(10);
+
+        $this->getActualVariant()->shouldReturn('variant test');
+        $this->getVariant()->shouldReturn($productVariant);
+        $this->getImagePath()->shouldReturn('/image/123/image.jpg');
+        $this->getQuantity()->shouldReturn(10);
+   }
+}
```
Comment (GracjanJozefczyk) [inline spec/Model/VariantPdfModelSpec.php:42]: Missing empty line

File: spec/Resolver/VariantImagePathResolverSpec.php [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+    function it_is_initializable(): void
+    {
+        $this->shouldHaveType(VariantImagePathResolver::class);
+    }
+
+    function it_resolve_empty_image_path(
+        ProductVariantInterface $variant,
+        ProductInterface $product,
+        Collection $productImages
+    ):  void
```
Comment (GracjanJozefczyk) [inline spec/Resolver/VariantImagePathResolverSpec.php:32]:
```suggestion
    ):  void {
```

File: spec/Resolver/VariantImagePathResolverSpec.php [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+
+        $this->resolve($variant, 'test:8000/')->shouldReturn('http://placehold.it/150x150');
+    }
+
+    function it_resolve_image_path(
+        ProductVariantInterface $variant,
+        ProductInterface $product,
+        Collection $productImages,
+        ProductImage $productImage
+    ):  void
```
Comment (GracjanJozefczyk) [inline spec/Resolver/VariantImagePathResolverSpec.php:46]:
```suggestion
    ):  void {
```

File: spec/Resolver/VariantImagePathResolverSpec.php [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+        $variant->getProduct()->willReturn($product);
+        $product->getImages()->willReturn($productImages);
+        $productImages->first()->willReturn($productImage);
+        $productImage->getPath()->willReturn('b4/c2/fc6b3202ee567e0fb05f293b709c.jpg');
+
+        $this
+            ->resolve($variant, 'http://127.0.0.1:8000')
+            ->shouldReturn('http://127.0.0.1:8000/media/image/b4/c2/fc6b3202ee567e0fb05f293b709c.jpg');
+    }
+}
```
Comment (GracjanJozefczyk) [inline spec/Resolver/VariantImagePathResolverSpec.php:58]: Missing empty line

File: src/Controller/Action/ExportWishlistToPdfAction.php [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+            $commandsArray->add($wishlistProductCommand);
+        }
+
+        $form = $this->formFactory->create(
+            WishlistCollectionType::class,
+            [
+                'items' => $commandsArray,
+            ],
+            [
+            'cart' => $cart,
```
Comment (GracjanJozefczyk) [inline src/Controller/Action/ExportWishlistToPdfAction.php:86]: Tab

File: src/Controller/Action/RemoveSelectedProductsFromWishlistAction.php [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+            $this->twigEnvironment->render('@BitBagSyliusWishlistPlugin/WishlistDetails/index.html.twig', [
+                'wishlist' => $wishlist,
+                'form' => $form->createView(),
+            ])
+        );
+    }
+
+    private function handleCartItems(array $wishlistProducts, Request $request): bool
+    {
+        $result = false;
```
Comment (GracjanJozefczyk) [inline src/Controller/Action/RemoveSelectedProductsFromWishlistAction.php:116]:
```suggestion
```

File: src/Controller/Action/RemoveSelectedProductsFromWishlistAction.php [hunk 30]
```diff
@@ -0,0 +1,142 @@
...
+                    if ($variant === $wishlistProductEntity->getVariant()) {
+                        $this->wishlistProductManager->remove($wishlistProductEntity);
+                    }
+                }
+            }
+        }
+
+        $this->wishlistProductManager->flush();
+
+        return $result;
```
Comment (GracjanJozefczyk) [inline src/Controller/Action/RemoveSelectedProductsFromWishlistAction.php:140]:
```suggestion
        return $result ?? false;
```

File: src/Form/Type/WishlistCollectionType.php [hunk 31]
```diff
@@ -0,0 +1,31 @@
+<?php
+
```
Comment (GracjanJozefczyk) [inline src/Form/Type/WishlistCollectionType.php:2]: Missing BitBag Header

File: src/Resolver/VariantImagePathResolver.php [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusWishlistPlugin\Resolver;
+
+use Sylius\Component\Core\Model\ProductVariantInterface;
+
+final class VariantImagePathResolver implements VariantImagePathResolverInterface
+{
+    public function resolve(ProductVariantInterface $variant, string $baseUrl)
```
Comment (GracjanJozefczyk) [inline src/Resolver/VariantImagePathResolver.php:17]: Missing return type

File: src/Resolver/VariantImagePathResolverInterface.php [hunk 33]
```diff
@@ -0,0 +1,18 @@
...
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusWishlistPlugin\Resolver;
+
+use Sylius\Component\Core\Model\ProductVariantInterface;
+
+interface VariantImagePathResolverInterface
+{
+    public function resolve(ProductVariantInterface $variant, string $baseUrl);
```
Comment (GracjanJozefczyk) [inline src/Resolver/VariantImagePathResolverInterface.php:17]: Return type

File: src/Resources/views/_wishlist_pdf.html.twig [hunk 34]
```diff
@@ -0,0 +1,44 @@
+{% import "@SyliusShop/Common/Macro/money.html.twig" as money %}
+<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="UTF-8">
+    <title>Title of the PDF</title>
```
Comment (GracjanJozefczyk) [inline src/Resources/views/_wishlist_pdf.html.twig:6]: Hard coded

File: tests/Behat/Context/Ui/WishlistContext.php [hunk 35]
```diff
@@ -38,22 +42,31 @@ final class WishlistContext implements Context
...
+        WishlistPageInterface        $wishlistPage,
         NotificationCheckerInterface $notificationChecker,
-        LoginerInterface $loginer,
-        WishlistCreatorInterface $wishlistCreator
-    ) {
+        LoginerInterface             $loginer,
+        WishlistCreatorInterface     $wishlistCreator,
+        Session                      $session,
+        RouterInterface              $router
+    )
```
Comment (GracjanJozefczyk) [inline tests/Behat/Context/Ui/WishlistContext.php:59]:
```suggestion
    ) {
```

File: tests/Behat/Context/Ui/WishlistContext.php [hunk 36]
```diff
@@ -156,6 +177,45 @@ public function iRemoveThisProduct(): void
...
+    public function iExportToPdfSelectedProductsFromWishlistAndFileIsDownloaded(): void
+    {
+        $this->wishlistPage->exportToPdfSelectedProductsFromWishlist();
+
+        $cookieName = $this->session->getName();
+        $sessionId = $this->session->getId();
+        $baseUrl = $this->getMinkParameter('base_url');
+        $domain = parse_url($baseUrl)['host'];
+
+        $cookieJar = \GuzzleHttp\Cookie\CookieJar::fromArray(array(
```
Comment (GracjanJozefczyk) [inline tests/Behat/Context/Ui/WishlistContext.php:200]:
```suggestion
        $cookieJar = \GuzzleHttp\Cookie\CookieJar::fromArray([
```

File: tests/Behat/Context/Ui/WishlistContext.php [hunk 37]
```diff
@@ -156,6 +177,45 @@ public function iRemoveThisProduct(): void
...
+            $cookieName => $sessionId,
+        ), $domain);
+
+        $guzzle = new \GuzzleHttp\Client(array(
```
Comment (GracjanJozefczyk) [inline tests/Behat/Context/Ui/WishlistContext.php:204]:
```suggestion
        $guzzle = new \GuzzleHttp\Client([
```

**Comment by Wojdylak** [comment]
This function is already done.

**Pull request closed without merging**